src/tools/pkcs11-tool.c

```diff
@@ -2491,8 +2492,13 @@ static void sign_data(CK_SLOT_ID slot, CK_SESSION_HANDLE session,
 		if ((getCLASS(session, key) == CKO_PRIVATE_KEY) && getALWAYS_AUTHENTICATE(session, key))
 			login(session,CKU_CONTEXT_SPECIFIC);
 
-		sig_len = sizeof(sig_buffer);
 		rv = p11->C_Sign(session, in_buffer, sz, sig_buffer, &sig_len);
+		if (rv == CKR_OK) {
+			sig_buffer = malloc(sig_len);
+			if (!sig_buffer)
+				util_fatal("malloc() failure\n");
+			rv = p11->C_Sign(session, in_buffer, sz, sig_buffer, &sig_len);
+		}
```

Comment:
I think this check should be written the other way round to fail in case this function call fails here (and to avoid needless nesting):
```suggestion
		if (rv != CKR_OK) {
			util_fatal("Failed to get signature length\n");
		}
		sig_buffer = malloc(sig_len);
		if (!sig_buffer) {
			util_fatal("malloc() failure\n");
		}
		rv = p11->C_Sign(session, in_buffer, sz, sig_buffer, &sig_len);
```

Reply:
same for the following block, which is similar.

Reply:
This patch changes the intent of the original code (before my proposed changes).  When the input is below a certain size threshold C_Sign is used.  If it fails a C_SignUpdate loop followed by C_SignFinal are attempted.  So the indentation is not actually unnecessary in the first section, unless you want to change the behavior.  My intention is only to make LMS signatures possible, not to alter any semantics.  The second section can be changed as you suggest though, so I'll do that.

Reply:
thank you for the clarification. I missed that part of the logic.